scalalib/src/mill/javalib/android/AndroidAppBundle.scala

```diff
@@ -0,0 +1,127 @@
+package mill.javalib.android
+
+import mill._
+import mill.scalalib._
+import mill.api.PathRef
+import mill.javalib.android.AndroidAppModule
+
+/**
+ * Defines a Trait for Android App Bundle Creation
+ */
+@mill.api.experimental
+trait AndroidAppBundle extends AndroidAppModule with JavaModule {
+
+  /**
+   * Adds `--proto-format` option to AAPT commands.
+   */
+  override def androidAaptOptions: T[Seq[String]] = Task {
+    super.androidAaptOptions() ++ Seq[String]("--proto-format")
+  }
+
+  /**
+   * Creates a zip with compiled resources and DEX files for the app bundle.
+   *
+   * @return PathRef to the bundle zip.
+   *
+   * For Bundle Zip Format please See the official
+   * [[https://developer.android.com/guide/app-bundle/app-bundle-format Documentation]]
+   */
+  def androidBundleZip: T[PathRef] = Task {
+    val dexFile = androidDex().path
+    val resFile = androidResources().path / "res.apk"
+    val baseDir = Task.dest / "base"
+    val appDir = Task.dest / "app"
+
+    // Copy dex and resource files into the bundle directory
+    os.copy(dexFile, Task.dest / dexFile.last)
+    os.unzip(resFile, appDir)
+    os.copy(
+      appDir / "AndroidManifest.xml",
+      baseDir / "manifest/AndroidManifest.xml",
+      createFolders = true
+    )
+
+    // Copy .pb files (compiled protobuf resources) and organize dex files using for loops
+    for (file <- os.walk(appDir).filter(_.ext == "pb")) {
+      os.copy(file, baseDir / file.last)
+    }
+
+    for ((file, idx) <- os.walk(Task.dest).filter(_.ext == "dex").zipWithIndex) {
+      os.copy(
+        file,
+        baseDir / "dex" / (if (idx == 0) "classes.dex" else s"classes${idx + 1}.dex"),
```

Comment:
Where is the documentation link for the `classes{n}.dex` file names?

Reply:
sorry forgot will add this time

Reply:
Such split is a part of multidex https://developer.android.com/build/multidex#mdex-on-l, you can reference this page if you want.

I'm not sure though why you don't preserve original file name: in your case original `classes2.dex` may become `classes3.dex` or `classes.dex`, for example, if the order of enumeration produced by `walk` is not the same as alphabetical ordering? Or are class files you are traversing have different naming?

I would suggest keeping original names, because `classes.dex` has a special treatment: it is always loaded first and some classes are packaged there specifically to allow application start.